### PR TITLE
ENH: add unit conversion derived signal from pcdsdevices

### LINKED ISSUE #894: Upstream UnitConversionDerivedSignal?
Unit-handling keeps coming up. It seems reasonable to add a dependency on one of the unit libraries to ophyd, given ophyd's role.

Should we upstream this an accept a unit dependency?

https://github.com/pcdshub/pcdsdevices/pull/518

### PR DESCRIPTION
**Pull request by klauer**
Summary
----------
* Fixes metadata keys on `FakeEpicsSignal`
* Fixes `put()` kwargs on `FakeEpicsSignal` to avoid frustratingly-common deprecation warnings
* Adds `UnitConversionDerivedSignal` with small test suite.

Closes #894 

Example
---------

This can be used as a device component like follows:

```python
    # component_name_to_convert_from = Cpt(Signal, value=0.)
    setpoint = Cpt(UnitConversionDerivedSignal,
                   derived_from='component_name_to_convert_from',
                   derived_units='s',
                   original_units='ns',
                   kind='hinted',
                   doc='Setpoint which handles the conversion.',
                   limits=(-10e-6, 10e-6),
                   )
```

What does this do?
* The above would use `self.component_name_to_convert_from` and proxy methods such as `.read()`, `.describe()`, and value/metadata subscriptions (as does the original `DerivedSignal` for those unaware).
* The instantiated signal would apply unit conversion from nanoseconds to seconds when reading from `setpoint`, and seconds to nanoseconds when writing back to `setpoint` (and then the original signal).
* This signal may have its own user-configurable limits. The default is to take the limits of the original signal (with unit conversion, of course)

This is in use in pcdsdevices currently.

### CODE REVIEW

**Comment by klauer** [comment]
Cycling after fixing CI in #980

**Comment by ZLLentz** [comment]
A few things:
- We've found this signal to be very useful at LCLS when the unit of the EPICS IOC is not the unit the user wants to record in their data stream. I've seen this code before so I don't have any line-by-line comments to add.
- This does add a `pint` dependency, which was discussed a fair bit in #894. If we add a dependency it should go in the `requirements.txt` file.
- What do you think about Tom's suggestion of dependency injection for the unit conversion?

**Comment by klauer** [comment]
> What do you think about Tom's suggestion of dependency injection for the unit conversion? #894 (comment)

I think it is premature to implement such a thing on a new feature, but am willing to accept Tom's patch in this PR for that functionality.

**Review by mrakitin — APPROVED** [review]
Looks good to me, and it's a pretty useful feature!

File: ophyd/sim.py [hunk 1]
```diff
@@ -1334,14 +1350,17 @@ class FakeEpicsSignalRO(SynSignalRO, FakeEpicsSignal):
...
 
 
 class FakeEpicsSignalWithRBV(FakeEpicsSignal):
     """
     FakeEpicsSignal with PV and PV_RBV; used in the AreaDetector PV naming
     scheme
     """
+
+    _metadata_keys = EpicsSignalWithRBV._metadata_keys
+
```
[the remark this answers is not in the document]
Comment (klauer) [reply ophyd/sim.py:1363]: I think I got that one as well, just above this class.

File: ophyd/units.py [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+    user_offset : any, optional
+        An optional user offset that will be *subtracted* when updating the
+        original signal, and *added* when calculating the derived value.
+        This offset should be supplied in ``derived_units`` and not
+        ``original_units``.
+
+        For example, if the original signal updates to a converted value of
+        500 ``derived_units`` and the ``user_offset`` is set to 100, this
+        ``DerivedSignal`` will show a value of 600.  When providing a new
+        setpoint, the ``user_offset`` will be subtracted.
```
[the remark this answers is not in the document]
Comment (klauer) [reply ophyd/units.py:74]: Could be useful, for sure. 👍

File: ophyd/units.py [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+        self.original_units = original_units
+        self._user_offset = user_offset
+        self._custom_limits = limits
+        super().__init__(derived_from, **kwargs)
+        self._metadata['units'] = derived_units
+
+        # Ensure that we include units in metadata callbacks, even if the
+        # original signal does not include them.
+        if 'units' not in self._metadata_keys:
+            self._metadata_keys = self._metadata_keys + ('units', )
```
[the remark this answers is not in the document]
Comment (klauer) [reply ophyd/units.py:114]: Tuple mutability feels a bit wrong, but I'm not opposed to accepting the suggestion. 🤷‍♂️

File: ophyd/units.py [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+
+    def forward(self, value):
+        '''Compute derived signal value -> original signal value'''
+        if self.user_offset is None:
+            raise ValueError(f'{self.name} must be set to a non-None value.')
```
[the remark this answers is not in the document]
Comment (klauer) [reply ophyd/units.py]: Good catch; resolved in c8943e4e

File: ophyd/sim.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mrakitin) [reply ophyd/sim.py:1363]: OK, I can see that now. Thanks.

**Review by mrakitin — APPROVED** [review]

**Review by ZLLentz — APPROVED** [review]

**Comment by mrakitin** [comment]
@klauer, ready to merge now. Do you mind resolving the merge conflicts, please?

**Comment by mrakitin** [comment]
Thanks, @klauer!

**Pull request merged**